### PR TITLE
feat(dashboard): Let users re-arrange native filters

### PR DESCRIPTION
**Pull request by m-ajay**
### SUMMARY
Allows users to re-arrange the dashboard native filters

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
ezgif com-gif-maker

### TESTING INSTRUCTIONS
 - Click Edit icon (✏️) in native filters section on the left side of the dashboard
 - Re-arrange the native filters in the opened modal box
 - Save and verify that the order has changed in the UI

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by jinghua-qa** [comment]
I have tested in the test environment, the drag and drop is not accurate, it always drop the filter to the bottom no matter i move up or move down, some time i just want to change the filter position from 1st to 2nd and it was dropped to the last

https://user-images.githubusercontent.com/81597121/128784760-b9f0ef16-7e2e-4cfa-a19c-706130be5f91.mov

**Comment by m-ajay** [comment]
@jinghua-qa Thanks for testing. It happens when we newly add filters and try re-arrange them. I have fixed it now.

File: superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FilterTabTitle.tsx [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+            role="button"
+            data-test="undo-button"
+            tabIndex={0}
+            onClick={() => restoreFilter(id)}
+          >
+            {t('Undo?')}
+          </StyledSpan>
+        )}
+      </StyledFilterTabTitle>
+      <div style={{ alignSelf: 'flex-end' }}>
```
Comment (michael-s-molina) [inline superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FilterTabTitle.tsx:180]: Avoid using `style`. Currently, we are using Emotion `styled` or `css` depending on the situation. You can check this [guide](https://github.com/apache/superset/issues/15264) for more information.

File: superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigModal.test.tsx [hunk 2]
```diff
@@ -30,6 +30,8 @@ import {
 import { render, screen, waitFor } from 'spec/helpers/testing-library';
 import mockDatasource, { id, datasourceId } from 'spec/fixtures/mockDatasource';
 import chartQueries from 'spec/fixtures/mockChartQueries';
+import { DndProvider } from 'react-dnd';
+import { HTML5Backend } from 'react-dnd-html5-backend';
```
Comment (michael-s-molina) [inline superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigModal.test.tsx:34]:
```suggestion
```

File: superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigModal.test.tsx [hunk 3]
```diff
@@ -149,10 +151,15 @@ beforeAll(() => {
...
-  });
+  return render(
+    <DndProvider backend={HTML5Backend}>
+      <FiltersConfigModal {...props} />
+    </DndProvider>,
+    {
+      useRedux: true,
+      initialState,
+    },
+  );
```
Comment (michael-s-molina) [inline superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigModal.test.tsx:162]:
```suggestion
  return render(<FiltersConfigModal {...props} />, {
    useRedux: true,
    useDnd: true,
    initialState,
  });
```

**Comment by michael-s-molina** [comment]
https://user-images.githubusercontent.com/70410625/129623092-a3da2276-f70b-4af3-be76-cacc4c3570dd.mov

- After the drop, the previous position is getting highlighted.
- The colors of selected and hover states could be more distinct. Currently, it's hard to see which item is hovered and which one is selected.
- In the drag animation, add padding to the text.

**Comment by michael-s-molina** [comment]
https://user-images.githubusercontent.com/70410625/129623964-edf862ba-0938-4658-9345-6959eb0dfb1d.mov

- Show the unsaved changes message if the user reorders a filter and presses cancel
- After canceling, the original order should be preserved. Currently, it's not canceling the reordering.

**Review by michael-s-molina — CHANGES_REQUESTED** [review]
@m-ajay Thanks for the contribution! I left some comments to improve even more the feature.

**Comment by junlincc** [comment]
it appears to me  that there's some lagging when reordering. could we optimize the perf? 
https://user-images.githubusercontent.com/67837651/130310816-c0157357-55cd-48a3-85ef-df292c97cdd4.mov

**Comment by m-ajay** [comment]
> it appears to me that there's some lagging when reordering. could we optimize the perf?
> https://user-images.githubusercontent.com/67837651/130310816-c0157357-55cd-48a3-85ef-df292c97cdd4.mov

@junlincc I would like to know whether it was a production build or development build.

File: superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FilterTabTitle.tsx [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+  padding-bottom: ${theme.gridUnit}px;
+  padding-left: ${2 * theme.gridUnit}px;
+  &:hover {
+    transition: all 0.3s;
+    span, .anticon {
+      color: ${theme.colors.primary.dark1};
+    }
+  }
+`}
+`;
```
Comment (michael-s-molina) [inline superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FilterTabTitle.tsx:49]:
```suggestion
const TabTitleContainer = styled.div<TabTitleContainerProps>`
  ${({ theme, isDragging }) => `
    display: flex;
    width: 100%;
    border-radius:  ${theme.borderRadius}px;
    opacity: ${isDragging ? 0.5 : 1};
    padding-top: ${theme.gridUnit}px;
    padding-bottom: ${theme.gridUnit}px;
    padding-left: ${2 * theme.gridUnit}px;
    &:hover {
      transition: all 0.3s;
      span, .anticon {
        color: ${theme.colors.primary.dark1};
      }
    }
  `}
`;
```

File: superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FilterTabs.tsx [hunk 5]
```diff
@@ -111,36 +71,42 @@ const FilterTabsContainer = styled(LineEditableTabs)`
...
         .ant-tabs-tab-remove > span {
           color: ${theme.colors.grayscale.base};
           transition: all 0.3s;
         }
+        .anticon {
+          color: ${theme.colors.grayscale.base};
+        }
+        :hover{
+          span {
+            color: ${theme.colors.grayscale.base} !important;
```
Comment (michael-s-molina) [inline superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FilterTabs.tsx:99]: We try to always avoid `!important`. Can you find another way to override the color? You can increase the specificity using `&`.

File: superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigModal.tsx [hunk 6]
```diff
@@ -208,7 +230,11 @@ export function FiltersConfigModal({
     if (values) {
       createHandleSave(
         filterConfigMap,
-        filterIds,
+        filterIds
+          .slice()
+          .sort(
+            (a, b) => orderedFilters.indexOf(a) - orderedFilters.indexOf(b),
+          ),
```
Comment (michael-s-molina) [inline superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigModal.tsx:237]: This code is duplicated in 3 places in this file. Could you extract it to a function?

**Comment by michael-s-molina** [comment]
@m-ajay Thanks for resolving the issues. The code looks great and I just left some final suggestions.

While testing I found this one:

https://user-images.githubusercontent.com/70410625/131578585-16f61e27-1f43-478e-a73a-9dd8a67f30f9.mov

**Comment by graceguo-supercat** [comment]
Thanks @michael-s-molina and @m-ajay I also played a little bit, here is my feedback:

Screen_Shot_2021-08-31_at_3_59_15_PM-2

1. only parent (top level filters) are shown in the filters list
2. In the edit mode, nested filters are also dragable and could be reordered. But there is no impact for display in the filter list, since nested filters don't show up in the filters list. 

I understand we want to allow user fully control the nested filter so we show them as first-class filter component. But make user move them up and down do not make sense to me. So my suggestion is:
- make child filter always next to parent filter,
- when user moves parent filter, children should be moved along,
- children (nested filters) itself, should not be dragable or rearrangeable.

@m-ajay @michael-s-molina @junlincc please let me know your opinion about this behavior.

**Comment by michael-s-molina** [comment]
@graceguo-supercat I agree with your points. We just need to make sure that the order is updated when setting the parent. 

We could also do something like this to clearly indicate to the user which filters are draggable and also to help with the visual group separation:

Screen Shot 2021-09-01 at 8 53 55 AM

It would be nice if the drag animation shows the whole group being dragged. If the group is big we could limit the animation to the first 3 and add `...`

**Comment by junlincc** [comment]
Let's discuss it during next Wednesday meeting. 🙏

**Comment by michael-s-molina** [comment]
@m-ajay I was just checking the latest version of the interface and I found many differences in relation to what we have on master. 

This is the master version:

Screen Shot 2021-10-04 at 3 01 57 PM

This is the latest version of this branch:

Screen Shot 2021-10-04 at 2 57 04 PM

I marked some of the differences:

- Double vertical bars on the left panel. The bar is appearing even if we have only one item.
- The Add Filter button is disappearing from the screen
- The top margin on the right panel is different
- We have a vertical scrollbar on the right panel even when the content fits
- The right border of the collapse panel is duplicated, giving the impression that is thicker than the other borders
- The selected filter state is different. Previously we had a background color indicating the selected filter
- The left padding on the left panel is different. Look how the filters are aligned with the Filters title on master

About the new feature:
- I noticed that the cursor is changing (grab/grabbing) only when I drag using the icon but if I drag using the name the cursor does not change
- The drag image contains a bottom padding
- The hover is changing the color of the text and delete icon but not the hover icon

These points are just something that I noticed very quickly but maybe we can have more. I suggest that you compare the proposed view with the one on master to make sure the behaviors are the same with the added drag and drop functionality.

Just ping me if any of these points are not clear or if you need any help 😉

**Comment by m-ajay** [comment]
re-arrange
@michael-s-molina Thanks for the comments. I have fixed the scrolling issue other style comments.

About
> The hover is changing the color of the text and delete icon but not the hover icon

The drag icon is not part of the filter title, but part of the filter group. That's why I did not change the color on hover.
Thanks!

File: superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FilterTitleContainer.tsx [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+                restoreFilter(id);
+              }}
+            >
+              {t('Undo?')}
+            </span>
+          )}
+        </div>
+        <div css={{ alignSelf: 'flex-end', marginLeft: 'auto' }}>
+          {isRemoved ? (
+            <></>
```
Comment (etr2460) [inline superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FilterTitleContainer.tsx:105]: why do we need an empty fragment? I assume `null` would be fine?

File: superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FilterTitleContainer.tsx [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+  ): React.ReactNode => {
+    const didAlreadyRender = rendered.indexOf(elementId) >= 0;
+    if (didAlreadyRender) {
+      return null;
+    }
+    let parent = null;
+    const element = nodeList.filter(el => el.id === elementId)[0];
+    if (!element) {
+      // eslint-disable-next-line no-console
+      console.warn(`Could not find filter with ID ${elementId}`);
```
Comment (etr2460) [inline superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FilterTitleContainer.tsx:132]: let's clean up the console logs before merging?

File: superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FilterTitlePane.tsx [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+  onRemove,
+  onRearrage,
+  restoreFilter,
+  currentFilterId,
+  filterGroups,
+  removedFilters,
+}) => {
+  const theme = useTheme();
+  return (
+    <>
```
Comment (etr2460) [inline superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FilterTitlePane.tsx:78]: fragment is unneeded

**Review by graceguo-supercat — APPROVED** [review]
LGTM

**Review by michael-s-molina — CHANGES_REQUESTED** [review]
Left some code suggestions and some problems that I found during the tests.

**Review by jinghua-qa — APPROVED** [review]
LGTM

**Review by michael-s-molina — APPROVED** [review]
LGTM. Thank you so much @m-ajay for your hard work and for addressing all the comments. This feature really benefits the user experience and it's a nice addition to the product. 👏🏼 🚀

**Pull request merged**